### DISCUSSION TITLE
UVK5 Roger Beep

### DISCUSSION THREAD

**Discussion by TexasAerospace** [Tunas1337/UV-K5-Modded-Firmwares#33] [Ideas]
From my understanding, the Roger beep is included in the firmware. Would it be possible to make the UVK5 use a different Roger beep?

**Comment by Tunas1337** [comment]
Technically, yes. However, we are still rather early in the reverse engineering process and no one knows how to do that.
What kind of roger beep do you have in mind?

**Reply by TexasAerospace** [reply]
Some sort of Motorola Beep, I’ll have to get back to you on the exact model.

**Reply by Sk3leCreeper** [reply]
A MotoTRBO beep would be interesting. The problem I could see with a lot of roger beeps, especially motorolas is that it's a sequence of tones, although the K5 does happen to have that fake MDC roger beep. Depends on how the beep is stored. This is the sequence for MotoTRBO:
![Untitled](https://github.com/Tunas1337/UV-K5-Modded-Firmwares/assets/74078464/2dc6f30c-d8f5-4c7a-be7a-6bbf1837f8be)

**Reply by Tunas1337** [reply]
Funnily enough, yes! This is possible. We reverse engineered both the regular two-tone roger and the MDC roger now, so we could try to replace one or the other. Idk if the two-tone beep would fit more than two tones but it's worth a try.

**Reply by Sk3leCreeper** [reply]
Awesome! I’m game for testing when a full firmware edit or mod for the uvmod_kitchen is ready.

**Reply by Tunas1337** [reply]
Best I could do is a mod to make the roger beep 1540Hz for 150ms and 1310Hz for 80ms. I'm still working on getting it pushed out, I'll edit this or comment when it's done. I also might record an example of what it would sound like.

EDIT: https://github.com/amnemonic/Quansheng_UV-K5_Firmware/pull/61/commits/96fb4eea452fdfb18db91dc840e9e84064d3dfa6
Example of what it sounds like: https://ttm.sh/BXR.mp3

**Reply by Sk3leCreeper** [reply]
I figured. Would adding more tone instructions break the firmware?

**Reply by Tunas1337** [reply]
Yes, unless you relocate everything below it.

**Comment by whosmatt** [comment]
> A MotoTRBO beep would be interesting. The problem I could see with a lot of roger beeps, especially motorolas is that it's a sequence of tones, although the K5 does happen to have that fake MDC roger beep. Depends on how the beep is stored. This is the sequence for MotoTRBO: ![Untitled](https://user-images.githubusercontent.com/74078464/252067164-2dc6f30c-d8f5-4c7a-be7a-6bbf1837f8be.png)

Just for reference, the MDC roger is not stored but actually generated using the FSK modem of the chip. The data it transmits is hardcoded and nonsensical though, it seems like a poor failed attempt at imitating the actual MDC databurst.

**Reply by Sk3leCreeper** [reply]
Oh yeah it's total nonsense, just sounds audio-wise similar. Interesting that it is actually generated on the fly each time, how is the regular 2-tone roger stored?

**Reply by Tunas1337** [reply]
> how is the regular 2-tone roger stored?

As a sequence of two instructions to the BK4819 to generate the tones: 500 Hz for 80 ms and 700 Hz for 80 ms.
![image](https://github.com/Tunas1337/UV-K5-Modded-Firmwares/assets/12097904/ab1767f2-dc15-4fa7-b883-ce0382e3446c)

**Reply by TexasAerospace** [reply]
Could we use this beep: https://youtu.be/2XyFVFjan1w?t=18 

as the roger beep? its a 3 tone.

**Comment by whosmatt** [comment]
the chip has tone generators that are used for roger, dtmf, 1750hz
